packages/contracts/contracts/SmartInvoiceFactoryV2.sol

```diff
@@ -0,0 +1,312 @@
+    mapping(bytes32 => uint256) public implementationsVersions;
+    mapping(address => bool) internal implementationExists;
+
+    event LogNewInvoice(
+        uint256 indexed index,
+        address invoice,
+        uint256[] amounts
+    );
+    event UpdateResolutionRate(
+        address indexed resolver,
+        uint256 indexed resolutionRate,
+        bytes32 details
+    );
+
+    address public immutable wrappedNativeToken;
+
+    constructor(address _wrappedNativeToken) {
+        require(
+            _wrappedNativeToken != address(0),
+            "invalid wrappedNativeToken"
+        );
+        wrappedNativeToken = _wrappedNativeToken;
+
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(ADMIN, msg.sender);
+    }
+
+    function _init(
+        address _invoiceAddress,
+        address _client,
+        address _provider,
+        uint8 _resolverType,
+        address _resolver,
+        uint256[] calldata _amounts,
+        bytes calldata _implementationData,
+        bytes32 _implementationType,
+        uint256 _implementationVersion,
+        address _implementationAddress
+    ) internal {
+        uint256 resolutionRate = resolutionRates[_resolver];
+        if (resolutionRate == 0) {
+            resolutionRate = 20;
+        }
+
+        ISmartInvoiceV2(_invoiceAddress).init(
+            _client,
+            _provider,
+            _encodeResolutionData(_resolverType, _resolver, resolutionRate),
+            _amounts,
+            wrappedNativeToken,
+            _implementationData,
+            _encodeImplementationData(
+                _implementationType,
+                _implementationVersion,
+                _implementationAddress,
+                invoiceCount
+            )
+        );
+
+        uint256 invoiceId = invoiceCount;
+        _invoices[invoiceId] = _invoiceAddress;
+        invoiceCount += 1;
+
+        emit LogNewInvoice(invoiceId, _invoiceAddress, _amounts);
+    }
+
+    // ******************
+    // Encoding
+    // ******************
+
+    function _encodeResolutionData(
+        uint8 _resolverType,
+        address _resolver,
+        uint256 _resolutionRate
+    ) internal pure returns (bytes memory) {
+        return abi.encode(_resolverType, _resolver, _resolutionRate);
+    }
+
+    function _encodeImplementationData(
+        bytes32 _implementationType,
+        uint256 _implementationVersion,
+        address _implementationAddress,
+        uint256 _invoiceId
+    ) internal pure returns (bytes memory) {
+        return
+            abi.encode(
+                _implementationType,
+                _implementationVersion,
+                _implementationAddress,
+                _invoiceId
+            );
+    }
+
+    // ******************
+    // Create
+    // ******************
+
+    function create(
+        address _client,
+        address _provider,
+        uint8 _resolverType,
+        address _resolver,
```

Comment:
_resolver and _resolverType should be encoded and included in the params as bytes instead. Makes it more arbitrary and flexible for some invoice types that include no arbitration or possibly more arbitration.

Reply:
I was thinking about this, since the `bytes` come from the client we can't rely on them to put these in there. So is it reasonable to abi encode these into the `bytes` to pass along?

Reply:
So since the frontend already have it structured, we can use ethers.js to encode all arbitration data (i.e. _resolver and _resolverType) and pass a `bytes calldata _resolverCtx` param in `create` function. Then in the `_init` function, only if `_resolverCtx` is NOT an empty `bytes`, we call `abi.encode(_resolverCtx, resolutionRate)` to get `resolutionData` and pass it in to `ISmartInovice(_invoiceAddress).init()`. Otherwise we pass an empty `bytes` as the resultionData param for `ISmartInovice(_invoiceAddress).init()`. The require statements in the `init` function for each implementation will decode and check if the resolver data is valid and all included so this should not increase security risks theoretically.

Reply:
ok, this is the flow and process for the implementationData at the moment, so we can bundle resolverData the same way.

Reply:
I created a new PR that avoids this issue. In PR #111, you decode and get the resolutionRate by calling back to the factory contract from within the escrow contract.